File: multiwalk.py
# import libraries
from subprocess import Popen
import subprocess
import os
import pandas as pd
import matplotlib.pyplot as plt

# input validation

# get the number of walkers for the simulation
while True:
    try:
        num_walkers = int(input("Enter the number of walkers (1 to 500): "))
        if 1 <= num_walkers <= 500:
            break
        else:
            print("Error: Please enter a number between 1 and 500")
    except ValueError:
        print("Error: Please enter an integer value")

# get the total number of steps each walker will take (max 50,000)
while True:
    try:
        num_steps = int(input("Enter the number of steps (2 to 50,000): "))
        if 2 <= num_steps <= 50000:
            break
        else:
            print("Error: Please enter a number between 1 and 50,000")
    except ValueError:
        print("Error: Please enter an integer value")

# get the x-axis drift
while True:
    try:
        driftX = float(input("Enter the x-axis drift (-1 to 1): "))
        if -1 <= driftX <= 1:
            break
        else:
            print("Error: Please enter a number between -1 and 1")
    except ValueError:
        print("Error: Please enter a value without a comma")

# get the y-axis drift
while True:
    try:
        driftY = float(input("Enter the y-axis drift (-1 to 1): "))
        if -1 <= driftY <= 1:
            break
        else:
            print("Error: Please enter a number between -1 and 1")
    except ValueError:
        print("Error: Please enter a value without a comma")

# get the side length of the square domain
while True:
    try:
        halfWidth = int(input("Enter half width of square domain (0 = unrestricted, max = 200): "))
        if 0 <= halfWidth <= 200:
            break
        else:
            print("Error: Please enter a number between 0 and 200")
    except ValueError:
        print("Error: Please enter an integer")

# get the diffusion type
while True:
    try:
        diffusionType = int(input("Enter the diffusion type (0 = normal, 1 = subdiffusion, 2 = superdiffusion): "))
        if 0 <= diffusionType <= 2:
            break
        else:
            print("Error: Please enter 0 for normal, 1 for subdiffusion, and 2 for superdiffusion")
    except ValueError:
        print("Error: Please enter a valid integer")


# Compile C++ code
# This is based on a snippet from BubbleSort.py in the Speed-Test guided project
try:
    subprocess.check_output("g++ -std=c++17 multiwalk.cpp", stdin=None, stderr=subprocess.STDOUT, shell=True)
except subprocess.CalledProcessError as e:
    print("error:\n", e.output)
    raise SystemExit

# run the C++ exe with command line argument
print("Opening C++ program")
args = f'a.exe {num_walkers} {num_steps} {driftX} {driftY} {halfWidth} {diffusionType}'

p = Popen(args, shell=True) # run the exe with num_walkers and num_steps
p.wait() # wait for the C++ program to finish

os.remove("a.exe") # remove the exe now that the program has created the CSV we need

df = pd.read_csv("randomwalk.csv") # read in CSV output from the C++ file
print("CSV read successfully") # notify user that the CSV was successfully read

# INDIVIDUAL WALKER PATHS
plt.figure(figsize=(14,12))
plt.subplot(2, 2, 1)

count_walkers = 0 # accumulator to count individual walkers
for i in range(2, len(df.columns), 2): # start at index 2, step by 2
    if count_walkers >= 5: # only doing the first 5 walkers to avoid overcrowding
        break
    x = df.columns[i] # x-axis
    y = df.columns[i+1] # y-axis
    walker_enum = x.split('_')[0] # splits walker0_x into a list ["walker0", "x"] and accesses only the first part
    plt.plot(df[x], df[y], label=walker_enum) # plot the x and y coordinates of each walker, label using walker_enum
    count_walkers += 1 # increment accumulator

plt.xlabel("X")
plt.ylabel("Y")
plt.title("Individual Random Walk Paths")
plt.grid(True)
plt.legend()

# HEATMAP OF VISITED LOCATIONS

plt.subplot(2, 2, 2)

x = [] # all x values
y = [] # all y values

# extend the lists so that we can store (in order) every coordinate that each walker has been to
for i in range(2, len(df.columns), 2): # skip first two columns and loop over walker's x values
    x.extend(df[df.columns[i]].values) # fetch all the x values for each walker
    y.extend(df[df.columns[i+1]].values) # fetch all the y values for each walker

plt.hist2d(x, y, bins=100) # 100 x 100 grid, counts/plots each matching (x,y) pair
plt.colorbar(label="Visit Count")
plt.xlabel("X")
plt.ylabel("Y")
plt.title("Heatmap of Visited Locations")
plt.grid(False)


# SQUARED DISPLACEMENT OVER TIME

plt.subplot(2, 2, 3) # 1 row, 2 columns, righthand subplot
plt.plot(df['step'], df['exp_squared_disp'])
plt.xlabel("Step")
plt.ylabel("Average Squared Displacement")
plt.title("Ensemble-Averaged Squared Displacement vs. Step")
plt.grid(True)

# HISTOGRAM OF WALKER DISPLACEMENTS

displacements = [] # collection of all final displacement numbers
for i in range(2, len(df.columns), 2): # skip first two cols, index through the x columns
    x_column = df.columns[i] # get the label of the ith column (x-axis)
    y_column = df.columns[i+1] # get the label of the i+1 column (y-axis)
    final_disp_x = df[x_column].iloc[-1] # get the last reported x coordinate
    final_disp_y = df[y_column].iloc[-1] # get the last reported y coordinate

    displacement = (final_disp_x**2 + final_disp_y**2)**0.5 # calculate the Euclidean distance from origin
    displacements.append(displacement) # append to the displacements list

plt.subplot(2, 2, 4)
plt.hist(displacements, bins=20, color='steelblue', edgecolor='black')
plt.xlabel("Displacement from Origin")
plt.ylabel("Number of Walkers")
plt.title("Histogram of Walker Displacements at Final Step")
plt.grid(True)

plt.tight_layout()
plt.show() # present the figures

os.remove("randomwalk.csv") # delete the csv file
print("CSV deleted")

File: multiwalk.cpp
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <cmath>
#include <algorithm>
using namespace std;

double computeSquaredDisplacement(const vector<vector<double>>& walkers) {
    // compute the expected squared displacement at each step
    double sumSquaredDisplacement = 0.0; // accumulator for the summation
    for (const auto& walker:walkers) {
        double squaredDisplacement = 0.0; // squared displacement for an individual walker
        for (int d = 0; d < 2; d++) {
            squaredDisplacement += walker[d] * walker[d]; // square the x and y components from position vector
        }
        sumSquaredDisplacement += squaredDisplacement; // add to the sum of the squared displacement
    }
    return sumSquaredDisplacement / walkers.size();
}


int main(int argc, char* argv[]) {

    // command line arguments from python frontend

    // get the number of walkers for the simulation
    int numWalkers;
    if (argc > 1) {
        numWalkers = stoi(argv[1]);
    } else {
        numWalkers = 10; // default number of walkers = 10
    }

    // number of steps
    int numSteps;
    if (argc > 2) {
        numSteps = stoi(argv[2]);
    } else {
        numSteps = 50; // default number of steps = 50
    }

    // x-axis drift
    double driftX = 0;
    if (argc > 3) {
        driftX = stod(argv[3]);
    }

    // y-axis drift
    double driftY = 0;
    if (argc > 4) {
        driftY = stod(argv[4]);
    }

    int domainSize = 100;
    if (argc > 5) {
        domainSize = stoi(argv[5]);
    }

    // diffusion type
    int diffusionType = 0;
    if (argc > 6) {
        diffusionType = stoi(argv[6]);
    }


    // random number generation
    // https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution.html
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> stepDistribution(-1.0, 1.0); // step -1, or 1
    uniform_real_distribution<double> positionDistribution(-domainSize, domainSize);
    uniform_real_distribution<double> paretoDraw(0.0, 1.0); // uniform distribution for pareto distribution

    // initialize walkers
    // vector storing each walker and their x,y coords initialized at (0, 0)
    vector<vector<double>> walkers(numWalkers, vector<double>(2, 0.0));

    // number of steps each walker must wait (subdiffusive), initialized at 0
    vector<int> waitCounter(numWalkers, 0);

    ofstream out("randomwalk.csv"); // initialize the output file
    out << "step,exp_squared_disp"; // step and expected squared displacement columns

    // add columns for the walkers
    for (int w = 0; w < numWalkers; w++ ) { // for every walker
        out << ",walker" << w << "_x,walker" << w << "_y"; // ex: walker0_x, walker0_y. x and y columns for each walker
    }
    out << "\n";


    // loop to add position data to the CSV for all steps and for all walkers
    for (int s = 0; s <= numSteps; s++) {

        // update the displacement
        double expectedSquaredDisplacement = computeSquaredDisplacement(walkers);

        // write the row to CSV
        out << s << "," << expectedSquaredDisplacement; // step, expected displacement
        for (auto& walker:walkers) {
            out << "," << walker[0] << "," << walker[1]; // x and y coordinates
        }
        out << "\n";

        // if user defined number of steps is reached, exit
        if (s == numSteps) {
            break;
        }

        for (int i = 0; i < numWalkers; i++) { // loop through each walker

            auto& walker = walkers[i]; // assign a walker to the current walker in the vector

            // pareto distribution time waiting (subdiffusion)
            if (diffusionType == 1) {
                if (waitCounter[i] > 0) { // if the walker has more than 0 steps to wait
                    waitCounter[i]--; // decrement number of steps to wait
                    continue; // walkers cant make two consecutive moves
                }

                double u = paretoDraw(gen); // draw a random number between 0 and 1
                double x = 10.0; // minimum waiting time
                double alpha = 0.5; // tail heaviness

                // inverse transformation sampling (citations #2 and #3)
                double wait = x / pow(1 - u, 1.0 / alpha);

                waitCounter[i] = round(wait); // round to accept an integer value
            }

            // normal step
            double xStep = stepDistribution(gen) + driftX; // random x-direction step with drift
            double yStep = stepDistribution(gen) + driftY; // random y-direction step with drift

            // superdiffusion
            if (diffusionType == 2) {
                uniform_real_distribution<double> angleDist(0, 2*3.14159); // choose a random angle between 0 and 2*pi

                // pareto distribution parameters
                double u = paretoDraw(gen);

                // set a cap on the uniform value to avoid blow ups
                if (u > 0.999999999) {
                    u = 0.999999999;
                }

                double xMin = 1.0; // minimum x value
                double alpha = 1.5; // tail index
                double radius = xMin / pow(1-u, 1.0/alpha); // inverse CDF of the pareto distribution

                // have to do polar coordinates since superdiffusion is tricky with grids
                double angle = angleDist(gen);
                xStep = radius*cos(angle);
                yStep = radius*sin(angle);
            }

            double newX = walker[0] + xStep; // updated x position
            double newY = walker[1] + yStep; // updated y position

            // square boundary detection
            if (domainSize != 0) {
                if (newX > domainSize || newX < -domainSize){
                    newX = walker[0];
                 }

                if (newY > domainSize || newY < -domainSize) {
                    newY = walker[1];
                }
            }

            walker[0] = newX;
            walker[1] = newY;
        }

    }
    return 0;
}
